spec/mempool/mempool.md

```diff
@@ -0,0 +1,152 @@
+
+Committed transactions are not in the mempool:  
+**INV2.** $\forall tx. \forall p \in Correct. \square(tx \in p.committed \implies tx \notin p.mempool)$
+
+In blockchain, a transaction is (or not) valid in a given state.
+That is, a transaction can be valid (or not) at a given height of the ledger.
+To model this, consider a transaction $tx$.
+Let $p.ledger.valid(tx)$ be such a check at the current height of the ledger at process $p$ (ABCI call).
+Our third invariant is that only valid transactions are present in the mempool:  
+**INV3.** $\forall tx, \forall p \in Correct. \square(tx \in p.mempool \implies p.ledger.valid(tx))$
+
+Finally, we require some progress from the mempool.
+Namely, if a transaction appears at a correct process then eventually it is committed or forever invalid.  
+**INV4** $\forall tx. \forall p \in Correct. \square(tx \in p.mempool \implies \lozenge\square(tx \in p.committed \vee \neg p.ledger.valid(tx)))$
+
+The above invariant ensures that if a transaction enters the mempool (at a correct process), then it eventually leaves it at a later time.
+For this to be true, the client application must ensure that the validity of a transaction converges toward some value.
+This meanss that there exists a height after which $valid(tx)$ always returns the same value.
+Such a requirement is termed _eventual non-oscillation_ in the [ABCI](https://github.com/cometbft/cometbft/blob/main/spec/abci/abci%2B%2B_app_requirements.md#mempool-connection-requirements) documentation.
+It also appears in [Ethereum](https://github.com/ethereum/go-ethereum/blob/5c51ef8527c47268628fe9be61522816a7f1b395/light/txpool.go#L401) as a transaction is always valid until a transaction from the same address executes with the same or higher nonce.
+A simple way to satisfy this for the programmer is by having $valid(tx)$ deterministic and stateless (e.g., a syntactic check).
+
+**Practical considerations.**
+In practice, as it requires to traverse the whole ledger, INV2 is too expensive.
+Instead, we would like to maintain this only over the last $\alpha$ committed transactions, for some parameter $\alpha$.
+Given a process $p$, we write $p.lcommitted$ the last $\alpha$ committed transactions at $p$.
+Invariant INV2 is replaced with:  
+**INV2a.** $\forall tx. \forall p \in Correct. \square(tx \in p.lcommitted \implies tx \notin p.mempool)$
+
+Another practical concern is with INV3.
+This invariant requires to have a green light from the client application before adding a transaction to the mempool.
+For efficiency, such a validation needs to be made at most $\beta$ times per transaction at each height, for some parameter $\beta$.
+Ideally, $\beta$ equals $1$.
+In practice, $\beta = f(T)$ for some function f of the maximal number of transactions T submitted between two heights.
+Given some transaction $tx$, variable $p.valid[tx]$ tracks the number of times the application was asked at the current height.
+Invariant INV3 is replaced with:  
```

Comment:
I don't understand why "replaced". In my mind they should be complementary. The transaction should be in the mempool if the previous check succeeded and the number of checks is limited.

Reply:
Thanks for point out this. The sentence was a bit unclear. It now reads: "A weaker version of INV3 is as follows: " 

INV3a is weaker than INV3 because we want to validate each transaction (at a given height) at most once. However, this requires to store possibly a very large map. Insted we use a bounded  (LRU) cache. 

Notice that INV2a and INV3 grasp the role of the `cache` variable in `clist_mempool.go`.